From 15a57e452a302bcd5a8cce512a16905bfab43778 Mon Sep 17 00:00:00 2001
From: dervoeti <lukas.krug@stackable.tech>
Date: Tue, 22 Jul 2025 09:52:55 +0200
Subject: Add CycloneDX plugin

---
 build.gradle | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/build.gradle b/build.gradle
index bb2e65ab..15a882a2 100644
--- a/build.gradle
+++ b/build.gradle
@@ -71,6 +71,7 @@ plugins {
     id 'eclipse'
     id "com.github.spotbugs" version "6.2.0"
     id "com.google.osdetector" version "1.7.3"
+    id "org.cyclonedx.bom" version "2.4.1"
 }
 
 allprojects {
@@ -87,6 +88,17 @@ apply from: 'gradle/formatting.gradle'
 licenseFile = rootProject.file('LICENSE.txt')
 noticeFile = rootProject.file('NOTICE.txt')
 
+cyclonedxBom {
+    includeConfigs = ["runtimeClasspath"]
+    includeLicenseText = false
+    skipConfigs = ["compileClasspath", "testCompileClasspath"]
+    projectType = "application"
+    schemaVersion = "1.6"
+    jsonOutput = file("build/reports/cyclonedx/bom.json")
+    outputFormat = "json"
+    componentVersion = opensearch_build
+}
+
 spotbugs {
     includeFilter = file('spotbugs-include.xml')
 }
