From 57c1974f0c25ea3e9d1f92dfebdc707c5ec56801 Mon Sep 17 00:00:00 2001
From: Benedikt Labrenz <benedikt@labrenz.org>
Date: Tue, 12 Aug 2025 14:59:12 +0200
Subject: Add CycloneDX plugin

---
 build.gradle | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/build.gradle b/build.gradle
index 6c09118..5401612 100644
--- a/build.gradle
+++ b/build.gradle
@@ -63,6 +63,21 @@ buildscript {
     }
 }
 
+plugins {
+    id "org.cyclonedx.bom" version "2.4.1"
+}
+
+cyclonedxBom {
+    includeConfigs = ["runtimeClasspath"]
+    includeLicenseText = false
+    skipConfigs = ["compileClasspath", "testCompileClasspath"]
+    projectType = "application"
+    schemaVersion = "1.6"
+    jsonOutput = file("build/reports/cyclonedx/bom.json")
+    outputFormat = "json"
+    componentVersion = versions.version
+}
+
 apply plugin: 'java'
 apply plugin: 'idea'
 apply plugin: 'opensearch.opensearchplugin'
@@ -340,4 +355,4 @@ task bwcTestSuite(type: RestIntegTestTask) {
 
 run {
     useCluster testClusters.integTest
-}
\ No newline at end of file
+}
