diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/http/ProfileServlet.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/http/ProfileServlet.java
index fc0ec7736ed8..e324ad6d49fd 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/http/ProfileServlet.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/http/ProfileServlet.java
@@ -76,6 +76,7 @@
  * Following event types are supported (default is 'cpu') (NOTE: not all OS'es support all events)
  * // Perf events:
  * //    cpu
+ * //    itimer
  * //    page-faults
  * //    context-switches
  * //    cycles
@@ -115,6 +116,7 @@ public class ProfileServlet extends HttpServlet {
   private enum Event {
 
     CPU("cpu"),
+    ITIMER("itimer"),
     ALLOC("alloc"),
     LOCK("lock"),
     PAGE_FAULTS("page-faults"),
