diff --git a/config/jvm.options b/config/jvm.options
index bcaa277..df9c34e 100644
--- a/config/jvm.options
+++ b/config/jvm.options
@@ -89,5 +89,5 @@
 ## OpenDistro Performance Analyzer
 -Dclk.tck=100
 -Djdk.attach.allowAttachSelf=true
--Djava.security.policy=/usr/share/opensearch/config/opensearch-performance-analyzer/opensearch_security.policy
+-Djava.security.policy=/stackable/opensearch/config/opensearch-performance-analyzer/opensearch_security.policy
 --add-opens=jdk.attach/sun.tools.attach=ALL-UNNAMED
diff --git a/opensearch-docker-entrypoint.sh b/opensearch-docker-entrypoint.sh
index d998663..e1795ce 100755
--- a/opensearch-docker-entrypoint.sh
+++ b/opensearch-docker-entrypoint.sh
@@ -9,7 +9,7 @@
 # in favor of a newly started container

 # Export OpenSearch Home
-export OPENSEARCH_HOME=/usr/share/opensearch
+export OPENSEARCH_HOME=/stackable/opensearch
 export OPENSEARCH_PATH_CONF=$OPENSEARCH_HOME/config

 # The virtual file /proc/self/cgroup should list the current cgroup
