From fd817e28e630f5fdd66e8fe5ba4dc4d2fb846e86 Mon Sep 17 00:00:00 2001
From: Benedikt Labrenz <benedikt@labrenz.org>
Date: Thu, 24 Jul 2025 14:39:31 +0200
Subject: Add-CycloneDX-plugin

---
 build.gradle | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/build.gradle b/build.gradle
index cb784cd..91a6a1e 100644
--- a/build.gradle
+++ b/build.gradle
@@ -54,7 +54,7 @@ plugins {
     id 'checkstyle'
     id 'org.ajoberstar.grgit' version '5.0.0'
     id 'org.gradle.test-retry' version '1.6.1'
-
+    id "org.cyclonedx.bom" version "2.3.1"
 }
 
 checkstyle {
@@ -121,6 +121,16 @@ apply plugin: 'opensearch.java-agent'
 licenseFile = rootProject.file('LICENSE.txt')
 noticeFile = rootProject.file('NOTICE.txt')
 
+cyclonedxBom {
+    includeConfigs = ["runtimeClasspath"]
+    includeLicenseText = false
+    skipConfigs = ["compileClasspath", "testCompileClasspath"]
+    projectType = "application"
+    schemaVersion = "1.6"
+    outputFormat = "json"
+    componentVersion = opensearch_build
+}
+
 spotless {
     java {
         licenseHeaderFile(file('license-header'))
