From f35a5e1087d0efeefbdf7c9a44a436e2e9f7429d Mon Sep 17 00:00:00 2001
From: Malte Sander <malte.sander.it@gmail.com>
Date: Thu, 25 Sep 2025 11:13:20 +0200
Subject: Include jackson dataformat xml dependency

---
 build.gradle               | 15 ++++++++++++++-
 gradle/dependencies.gradle |  1 +
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/build.gradle b/build.gradle
index e6c0449952..c1aa6e2f55 100644
--- a/build.gradle
+++ b/build.gradle
@@ -180,13 +180,15 @@ ext {
     libs.slf4jLog4j2,
     libs.junitPlatformLanucher,
     libs.jacksonDatabindYaml,
+    libs.jacksonDatabindXml,
     project(":test-common:test-common-util")
   ]
 
   log4jReleaseLibs = [
     libs.slf4jLog4j2,
     libs.log4j1Bridge2Api,
-    libs.jacksonDatabindYaml
+    libs.jacksonDatabindYaml,
+    libs.jacksonDatabindXml
   ]
 
   log4j2Libs = [
@@ -1066,6 +1068,7 @@ project(':core') {
     implementation libs.jacksonDataformatCsv
     implementation libs.jacksonJDK8Datatypes
     implementation libs.jacksonDatabindYaml
+    implementation libs.jacksonDatabindXml
     implementation libs.joptSimple
     implementation libs.jose4j
     implementation libs.metrics
@@ -1484,6 +1487,7 @@ project(':group-coordinator') {
     testImplementation project(':server-common').sourceSets.test.output
     testImplementation project(':coordinator-common').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation testLog4j2Libs
@@ -1606,6 +1610,7 @@ project(':test-common:test-common-runtime') {
     implementation libs.junitPlatformLanucher
     implementation libs.junitJupiter
     implementation libs.jacksonDatabindYaml
+    implementation libs.jacksonDatabindXml
     implementation libs.slf4jApi
 
     testImplementation libs.junitJupiter
@@ -2079,6 +2084,7 @@ project(':raft') {
     testImplementation project(':clients')
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.jqwik
@@ -2176,6 +2182,7 @@ project(':server-common') {
     testImplementation project(':clients')
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation testLog4j2Libs
@@ -2317,6 +2324,7 @@ project(':storage') {
     testImplementation project(':transaction-coordinator')
     testImplementation libs.hamcrest
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.bcpkix
@@ -2634,6 +2642,7 @@ project(':shell') {
     testImplementation project(':server-common')
     testImplementation project(':server-common').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation testLog4j2Libs
 
@@ -2683,6 +2692,7 @@ project(':streams') {
 
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.bcpkix
     testImplementation libs.hamcrest
@@ -2828,6 +2838,7 @@ project(':streams:streams-scala') {
     testImplementation project(':streams:test-utils')
 
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoJunitJupiter // supports MockitoExtension
     testImplementation testLog4j2Libs
@@ -2932,6 +2943,7 @@ project(':streams:test-utils') {
 
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.hamcrest
@@ -3610,6 +3622,7 @@ project(':connect:runtime') {
     testImplementation project(':server-common').sourceSets.test.output
 
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.mockitoJunitJupiter
diff --git a/gradle/dependencies.gradle b/gradle/dependencies.gradle
index 10173814e5..1a8c1bac15 100644
--- a/gradle/dependencies.gradle
+++ b/gradle/dependencies.gradle
@@ -156,6 +156,7 @@ libs += [
   jacksonAnnotations: "com.fasterxml.jackson.core:jackson-annotations:$versions.jackson",
   jacksonDatabind: "com.fasterxml.jackson.core:jackson-databind:$versions.jackson",
   jacksonDatabindYaml: "com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:$versions.jackson",
+  jacksonDatabindXml: "com.fasterxml.jackson.dataformat:jackson-dataformat-xml:$versions.jackson",
   jacksonDataformatCsv: "com.fasterxml.jackson.dataformat:jackson-dataformat-csv:$versions.jackson",
   jacksonJDK8Datatypes: "com.fasterxml.jackson.datatype:jackson-datatype-jdk8:$versions.jackson",
   jacksonBlackbird: "com.fasterxml.jackson.module:jackson-module-blackbird:$versions.jackson",
