From e69a23a6b3181582d0638db00cf1caf81eff56ff Mon Sep 17 00:00:00 2001
From: Andrew Kenworthy <andrew.kenworthy@stackable.tech>
Date: Thu, 16 Oct 2025 09:24:39 +0200
Subject: re-parse flow

---
 .../nifi/authorization/FileAccessPolicyProvider.java     | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/nifi-framework-bundle/nifi-framework/nifi-file-authorizer/src/main/java/org/apache/nifi/authorization/FileAccessPolicyProvider.java b/nifi-framework-bundle/nifi-framework/nifi-file-authorizer/src/main/java/org/apache/nifi/authorization/FileAccessPolicyProvider.java
index cb16cd1a30..17b3f3929d 100644
--- a/nifi-framework-bundle/nifi-framework/nifi-file-authorizer/src/main/java/org/apache/nifi/authorization/FileAccessPolicyProvider.java
+++ b/nifi-framework-bundle/nifi-framework/nifi-file-authorizer/src/main/java/org/apache/nifi/authorization/FileAccessPolicyProvider.java
@@ -754,7 +754,12 @@ public class FileAccessPolicyProvider implements ConfigurableAccessPolicyProvide
      * then use "root" as a placeholder.
      */
     public void replaceWithRootGroupId() throws JAXBException {
+        if (rootGroupId == null) {
+            logger.info("Parsing flow as rootGroupId is not yet defined");
+            parseFlow();
+        }
         if (rootGroupId != null) {
+            logger.info("Parsing root group with {}", rootGroupId);
             Authorizations authorizations = this.authorizationsHolder.get().getAuthorizations();
             boolean authorizationsChanged = false;
             for (Policy policy: authorizations.getPolicies().getPolicy()) {
@@ -769,6 +774,10 @@ public class FileAccessPolicyProvider implements ConfigurableAccessPolicyProvide
             if (authorizationsChanged) {
                 saveAuthorizations(authorizations);
             }
+        } else {
+            // this is not expected as this is called from the flow service
+            // once it has been configured
+            logger.info("rootGroupId still not established!");
         }
     }
 
