From f77fdcc430c86388d9c575b8351488691915876b Mon Sep 17 00:00:00 2001
From: Razvan-Daniel Mihai <84674+razvan@users.noreply.github.com>
Date: Wed, 13 May 2026 17:25:07 +0300
Subject: Include jackson dataformat xml dependency

---
 build.gradle               | 15 ++++++++++++++-
 gradle/dependencies.gradle |  1 +
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/build.gradle b/build.gradle
index 2bdfb80d85..0ed19d18bb 100644
--- a/build.gradle
+++ b/build.gradle
@@ -185,13 +185,15 @@ ext {
     libs.slf4jLog4j2,
     libs.junitPlatformLanucher,
     libs.jacksonDatabindYaml,
+    libs.jacksonDatabindXml,
     project(":test-common:test-common-util")
   ]
 
   log4jReleaseLibs = [
     libs.slf4jLog4j2,
     libs.log4j1Bridge2Api,
-    libs.jacksonDatabindYaml
+    libs.jacksonDatabindYaml,
+    libs.jacksonDatabindXml
   ]
 
   log4j2Libs = [
@@ -1120,6 +1122,7 @@ project(':core') {
     implementation libs.jacksonDataformatCsv
     implementation libs.jacksonJDK8Datatypes
     implementation libs.jacksonDatabindYaml
+    implementation libs.jacksonDatabindXml
     implementation libs.joptSimple
     implementation libs.jose4j
     implementation libs.metrics
@@ -1534,6 +1537,7 @@ project(':group-coordinator') {
     testImplementation project(':server-common').sourceSets.test.output
     testImplementation project(':coordinator-common').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation testLog4j2Libs
@@ -1656,6 +1660,7 @@ project(':test-common:test-common-runtime') {
     implementation libs.junitPlatformLanucher
     implementation libs.junitJupiter
     implementation libs.jacksonDatabindYaml
+    implementation libs.jacksonDatabindXml
     implementation libs.slf4jApi
 
     testImplementation libs.junitJupiter
@@ -2130,6 +2135,7 @@ project(':raft') {
     testImplementation project(':clients')
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.jqwik
@@ -2227,6 +2233,7 @@ project(':server-common') {
     testImplementation project(':clients')
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation testLog4j2Libs
@@ -2368,6 +2375,7 @@ project(':storage') {
     testImplementation project(':transaction-coordinator')
     testImplementation libs.hamcrest
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.bcpkix
@@ -2685,6 +2693,7 @@ project(':shell') {
     testImplementation project(':server-common')
     testImplementation project(':server-common').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation testLog4j2Libs
 
@@ -2734,6 +2743,7 @@ project(':streams') {
 
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.bcpkix
     testImplementation libs.hamcrest
@@ -2879,6 +2889,7 @@ project(':streams:streams-scala') {
     testImplementation project(':streams:test-utils')
 
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoJunitJupiter // supports MockitoExtension
     testImplementation testLog4j2Libs
@@ -2983,6 +2994,7 @@ project(':streams:test-utils') {
 
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.hamcrest
@@ -3678,6 +3690,7 @@ project(':connect:runtime') {
     testImplementation project(':server-common').sourceSets.test.output
 
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.mockitoJunitJupiter
diff --git a/gradle/dependencies.gradle b/gradle/dependencies.gradle
index fb0465b66e..094d4545be 100644
--- a/gradle/dependencies.gradle
+++ b/gradle/dependencies.gradle
@@ -157,6 +157,7 @@ libs += [
   jacksonAnnotations: "com.fasterxml.jackson.core:jackson-annotations:$versions.jackson",
   jacksonDatabind: "com.fasterxml.jackson.core:jackson-databind:$versions.jackson",
   jacksonDatabindYaml: "com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:$versions.jackson",
+  jacksonDatabindXml: "com.fasterxml.jackson.dataformat:jackson-dataformat-xml:$versions.jackson",
   jacksonDataformatCsv: "com.fasterxml.jackson.dataformat:jackson-dataformat-csv:$versions.jackson",
   jacksonJDK8Datatypes: "com.fasterxml.jackson.datatype:jackson-datatype-jdk8:$versions.jackson",
   jacksonBlackbird: "com.fasterxml.jackson.module:jackson-module-blackbird:$versions.jackson",
