From 9af5f63f8624d3744079ba032ab105b4f256cb68 Mon Sep 17 00:00:00 2001
From: dervoeti <lukas.krug@stackable.tech>
Date: Tue, 22 Jul 2025 09:52:55 +0200
Subject: Add CycloneDX plugin

---
 build.gradle | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/build.gradle b/build.gradle
index af480ede..73e27030 100644
--- a/build.gradle
+++ b/build.gradle
@@ -69,6 +69,7 @@ plugins {
     id 'eclipse'
     id "com.github.spotbugs" version "6.4.4"
     id "com.google.osdetector" version "1.7.3"
+    id "org.cyclonedx.bom" version "2.3.1"
 }
 
 allprojects {
@@ -85,6 +86,16 @@ apply from: 'gradle/formatting.gradle'
 licenseFile = rootProject.file('LICENSE.txt')
 noticeFile = rootProject.file('NOTICE.txt')
 
+cyclonedxBom {
+    includeConfigs = ["runtimeClasspath"]
+    includeLicenseText = false
+    skipConfigs = ["compileClasspath", "testCompileClasspath"]
+    projectType = "application"
+    schemaVersion = "1.6"
+    outputFormat = "json"
+    componentVersion = opensearch_build
+}
+
 spotbugs {
     includeFilter = file('spotbugs-include.xml')
 }
