From a91d18e7227aea84fb0baaafc8d253d321a46a0b Mon Sep 17 00:00:00 2001
From: Lars Francke <git@lars-francke.de>
Date: Thu, 12 Dec 2024 10:09:47 +0100
Subject: Change Gradle to use the Nexus Build Repo

---
 build.gradle | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/build.gradle b/build.gradle
index fb0015e49a..7497ce70a3 100644
--- a/build.gradle
+++ b/build.gradle
@@ -20,7 +20,9 @@ import java.nio.charset.StandardCharsets
 
 buildscript {
   repositories {
-    mavenCentral()
+    maven {
+      url 'https://build-repo.stackable.tech/repository/maven-public/'
+    }
   }
   apply from: "$rootDir/gradle/dependencies.gradle"
 
@@ -208,7 +210,9 @@ ext {
 allprojects {
 
   repositories {
-    mavenCentral()
+    maven {
+      url 'https://build-repo.stackable.tech/repository/maven-public/'
+    }
   }
 
   dependencyUpdates {
