From f9ae1b0f4f9f1cf1beaa4ce589e2af9fde344b5b Mon Sep 17 00:00:00 2001
From: Malte Sander <malte.sander.it@gmail.com>
Date: Tue, 23 Sep 2025 09:55:40 +0200
Subject: Include jackson dataformat xml dependency

---
 build.gradle               | 13 +++++++++++++
 gradle/dependencies.gradle |  1 +
 2 files changed, 14 insertions(+)

diff --git a/build.gradle b/build.gradle
index 7497ce70a3..021adcb3db 100644
--- a/build.gradle
+++ b/build.gradle
@@ -185,12 +185,14 @@ ext {
     libs.slf4jLog4j2,
     libs.junitPlatformLanucher,
     libs.jacksonDatabindYaml,
+    libs.jacksonDatabindXml,
     project(":test-common:test-common-util")
   ]
 
   log4jReleaseLibs = [
     libs.slf4jLog4j2,
     libs.log4j1Bridge2Api,
+    libs.jacksonDatabindXml,
     libs.jacksonDatabindYaml
   ]
 
@@ -1134,6 +1136,7 @@ project(':core') {
     implementation libs.jacksonDataformatCsv
     implementation libs.jacksonJDK8Datatypes
     implementation libs.jacksonDatabindYaml
+    implementation libs.jacksonDatabindXml
     implementation libs.joptSimple
     implementation libs.jose4j
     implementation libs.metrics
@@ -1534,6 +1537,7 @@ project(':group-coordinator') {
     testImplementation project(':server-common').sourceSets.test.output
     testImplementation project(':coordinator-common').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation testLog4j2Libs
@@ -1654,6 +1658,7 @@ project(':test-common:test-common-runtime') {
     implementation libs.junitPlatformLanucher
     implementation libs.junitJupiter
     implementation libs.jacksonDatabindYaml
+    implementation libs.jacksonDatabindXml
     implementation libs.slf4jApi
 
     testImplementation libs.junitJupiter
@@ -2086,6 +2091,7 @@ project(':raft') {
     testImplementation project(':clients')
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.jqwik
@@ -2183,6 +2189,7 @@ project(':server-common') {
     testImplementation project(':clients')
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation testLog4j2Libs
@@ -2320,6 +2327,7 @@ project(':storage') {
     testImplementation project(':server-common').sourceSets.test.output
     testImplementation libs.hamcrest
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.bcpkix
@@ -2636,6 +2644,7 @@ project(':shell') {
     testImplementation project(':server-common')
     testImplementation project(':server-common').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation testLog4j2Libs
 
@@ -2685,6 +2694,7 @@ project(':streams') {
 
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.bcpkix
     testImplementation libs.hamcrest
@@ -2829,6 +2839,7 @@ project(':streams:streams-scala') {
     testImplementation project(':streams:test-utils')
 
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoJunitJupiter // supports MockitoExtension
     testImplementation testLog4j2Libs
@@ -2932,6 +2943,7 @@ project(':streams:test-utils') {
 
     testImplementation project(':clients').sourceSets.test.output
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.hamcrest
@@ -3569,6 +3581,7 @@ project(':connect:runtime') {
     testImplementation project(':server-common').sourceSets.test.output
 
     testImplementation libs.jacksonDatabindYaml
+    testImplementation libs.jacksonDatabindXml
     testImplementation libs.junitJupiter
     testImplementation libs.mockitoCore
     testImplementation libs.mockitoJunitJupiter
diff --git a/gradle/dependencies.gradle b/gradle/dependencies.gradle
index 66eca369aa..06aad64e6c 100644
--- a/gradle/dependencies.gradle
+++ b/gradle/dependencies.gradle
@@ -152,6 +152,7 @@ libs += [
   jacksonAnnotations: "com.fasterxml.jackson.core:jackson-annotations:$versions.jackson",
   jacksonDatabind: "com.fasterxml.jackson.core:jackson-databind:$versions.jackson",
   jacksonDatabindYaml: "com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:$versions.jackson",
+  jacksonDatabindXml: "com.fasterxml.jackson.dataformat:jackson-dataformat-xml:$versions.jackson",
   jacksonDataformatCsv: "com.fasterxml.jackson.dataformat:jackson-dataformat-csv:$versions.jackson",
   jacksonJDK8Datatypes: "com.fasterxml.jackson.datatype:jackson-datatype-jdk8:$versions.jackson",
   jacksonBlackbird: "com.fasterxml.jackson.module:jackson-module-blackbird:$versions.jackson",
