From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Larsen <nick.larsen@stackable.tech>
Date: Wed, 29 Jan 2025 16:50:30 +0100
Subject: Add cyclonedx-plugin

---
 build.gradle | 42 ++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 42 insertions(+)

diff --git a/build.gradle b/build.gradle
index 5b064f4203..2d7b442e7b 100644
--- a/build.gradle
+++ b/build.gradle
@@ -51,6 +51,48 @@ plugins {
   //  We are going to drop JDK8 support. Hence, the spotless is upgrade to newest version and be applied only if the build env is compatible with JDK 11.
   //  spotless 6.15.0+ has issue in runtime with JDK8 even through we define it with `apply:false`. see https://github.com/diffplug/spotless/issues/2156 for more details
   id 'com.diffplug.spotless' version "6.14.0" apply false
+  id 'org.cyclonedx.bom' version '1.10.0'
+}
+
+cyclonedxBom {
+    // Specified the type of project being built. Defaults to 'library'
+    projectType = "application"
+    // Specified the version of the CycloneDX specification to use. Defaults to '1.5'
+    schemaVersion = "1.5"
+    // Boms destination directory. Defaults to 'build/reports'
+    destination = file("build/reports")
+    // The file name for the generated BOMs (before the file format suffix). Defaults to 'bom'
+    outputName = "bom"
+    // The file format generated, can be xml, json or all for generating both. Defaults to 'all'
+    outputFormat = "json"
+    includeConfigs = ["runtimeClasspath"]
+    // Exclude test components. This list needs to be checked and, if it changed, updated for every new Kafka version.
+    // The list can be obtained by running `gradle projects | grep upgrade-system-tests`
+    skipProjects = [
+      'upgrade-system-tests-0100',
+      'upgrade-system-tests-0101',
+      'upgrade-system-tests-0102',
+      'upgrade-system-tests-0110',
+      'upgrade-system-tests-10',
+      'upgrade-system-tests-11',
+      'upgrade-system-tests-20',
+      'upgrade-system-tests-21',
+      'upgrade-system-tests-22',
+      'upgrade-system-tests-23',
+      'upgrade-system-tests-24',
+      'upgrade-system-tests-25',
+      'upgrade-system-tests-26',
+      'upgrade-system-tests-27',
+      'upgrade-system-tests-28',
+      'upgrade-system-tests-30',
+      'upgrade-system-tests-31',
+      'upgrade-system-tests-32',
+      'upgrade-system-tests-33',
+      'upgrade-system-tests-34',
+      'upgrade-system-tests-35',
+      'upgrade-system-tests-36',
+      'upgrade-system-tests-37'
+    ]
 }
 
 ext {

base-commit: 84caaa6e9da06435411510a81fa321d4f99c351f
-- 
2.40.1

