From e5102449fe825cfbba20ce6ace1f51cd91550780 Mon Sep 17 00:00:00 2001
From: Lars Francke <git@lars-francke.de>
Date: Thu, 12 Dec 2024 10:09:47 +0100
Subject: [PATCH] Change Gradle to use the Nexus Build Repo

---
 build.gradle | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/build.gradle b/build.gradle
index 92082fe7cf..3b56a2ad98 100644
--- a/build.gradle
+++ b/build.gradle
@@ -20,7 +20,9 @@ import java.nio.charset.StandardCharsets
 
 buildscript {
   repositories {
-    mavenCentral()
+    maven {
+      url 'https://build-repo.stackable.tech/repository/maven-public/'
+    }
   }
   apply from: "$rootDir/gradle/dependencies.gradle"
 
@@ -126,7 +128,9 @@ ext {
 allprojects {
 
   repositories {
-    mavenCentral()
+    maven {
+      url 'https://build-repo.stackable.tech/repository/maven-public/'
+    }
   }
 
   dependencyUpdates {
-- 
2.47.1

