From 43ecc157fdef2ff6937ddbede5321d82730fa37b Mon Sep 17 00:00:00 2001
From: dervoeti <lukas.krug@stackable.tech>
Date: Fri, 29 Aug 2025 18:55:48 +0200
Subject: testing

---
 airflow-core/src/airflow/settings.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/airflow-core/src/airflow/settings.py b/airflow-core/src/airflow/settings.py
index ce36f1e676..ef68f185be 100644
--- a/airflow-core/src/airflow/settings.py
+++ b/airflow-core/src/airflow/settings.py
@@ -75,6 +75,7 @@ HEADER = "\n".join(
         r"____  /| |_  /__  ___/_  /_ __  /_  __ \_ | /| / /",
         r"___  ___ |  / _  /   _  __/ _  / / /_/ /_ |/ |/ /",
         r" _/_/  |_/_/  /_/    /_/    /_/  \____/____/|__/",
+        r"modified by Stackable!",
     ]
 )
 
