From b1f266eba4ae63ba23c254193f176362a2aafa52 Mon Sep 17 00:00:00 2001
From: Sebastian Bernauer <sebastian.bernauer@stackable.tech>
Date: Tue, 27 Jan 2026 13:16:16 +0100
Subject: Support configuring S3 endpoint and path style access in Iceberg S3

---
 .../iceberg/aws/S3IcebergFileIOProvider.java  | 29 ++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-aws/src/main/java/org/apache/nifi/services/iceberg/aws/S3IcebergFileIOProvider.java b/nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-aws/src/main/java/org/apache/nifi/services/iceberg/aws/S3IcebergFileIOProvider.java
index 6581993c22..61b73cba39 100644
--- a/nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-aws/src/main/java/org/apache/nifi/services/iceberg/aws/S3IcebergFileIOProvider.java
+++ b/nifi-extension-bundles/nifi-iceberg-bundle/nifi-iceberg-aws/src/main/java/org/apache/nifi/services/iceberg/aws/S3IcebergFileIOProvider.java
@@ -102,12 +102,32 @@ public class S3IcebergFileIOProvider extends AbstractControllerService implement
             )
             .build();
 
+    static final PropertyDescriptor ENDPOINT_OVERRIDE = new PropertyDescriptor.Builder()
+            .name("Endpoint Override URL")
+            .description("Endpoint URL to use instead of the AWS default including scheme, host, port, and path. " +
+                    "The AWS libraries select an endpoint URL based on the AWS region, but this property overrides " +
+                    "the selected endpoint URL, allowing use with other S3-compatible endpoints.")
+            .expressionLanguageSupported(ExpressionLanguageScope.ENVIRONMENT)
+            .required(false)
+            .addValidator(StandardValidators.URL_VALIDATOR)
+            .build();
+
+    static final PropertyDescriptor USE_PATH_STYLE_ACCESS = new PropertyDescriptor.Builder()
+            .name("Use Path Style Access")
+            .description("Path-style access can be enforced by setting this property to true. Set it to true if your endpoint does not support " +
+                    "virtual-hosted-style requests, only path-style requests.")
+            .allowableValues("true", "false")
+            .defaultValue(String.valueOf(S3FileIOProperties.PATH_STYLE_ACCESS_DEFAULT))
+            .build();
+
     private static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS = List.of(
             AUTHENTICATION_STRATEGY,
             ACCESS_KEY_ID,
             SECRET_ACCESS_KEY,
             SESSION_TOKEN,
-            CLIENT_REGION
+            CLIENT_REGION,
+            ENDPOINT_OVERRIDE,
+            USE_PATH_STYLE_ACCESS
     );
 
     private final Map<String, String> standardProperties = new ConcurrentHashMap<>();
@@ -159,6 +179,13 @@ public class S3IcebergFileIOProvider extends AbstractControllerService implement
             }
         }
 
+        if (context.getProperty(ENDPOINT_OVERRIDE).isSet()) {
+            final String endpoint = context.getProperty(ENDPOINT_OVERRIDE).evaluateAttributeExpressions().getValue();
+            contextProperties.put(S3FileIOProperties.ENDPOINT, endpoint);
+        }
+        final String pathStyleAccess = context.getProperty(USE_PATH_STYLE_ACCESS).getValue();
+        contextProperties.put(S3FileIOProperties.PATH_STYLE_ACCESS, pathStyleAccess);
+
         // HttpURLConnection Client Type avoids additional dependencies
         contextProperties.put(HttpClientProperties.CLIENT_TYPE, HttpClientProperties.CLIENT_TYPE_URLCONNECTION);
 
