From b03f930266c43347645b4e978fa97cf644b270f8 Mon Sep 17 00:00:00 2001
From: Lars Francke <git@lars-francke.de>
Date: Thu, 12 Dec 2024 10:09:47 +0100
Subject: Change Gradle to use the Nexus Build Repo

---
 build.gradle | 44 +++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 43 insertions(+), 1 deletion(-)

diff --git a/build.gradle b/build.gradle
index 8b44b09087..fadb64fda5 100644
--- a/build.gradle
+++ b/build.gradle
@@ -48,6 +48,48 @@ plugins {
   // artifacts - see https://github.com/johnrengelman/shadow/issues/901
   id 'com.github.johnrengelman.shadow' version '8.1.0' apply false
   id 'com.diffplug.spotless' version '6.14.0' apply false // 6.14.1 and newer require Java 11 at compile time, so we can't upgrade until AK 4.0
+  id 'org.cyclonedx.bom' version '1.10.0'
+}
+
+cyclonedxBom {
+    // Specified the type of project being built. Defaults to 'library'
+    projectType = "application"
+    // Specified the version of the CycloneDX specification to use. Defaults to '1.5'
+    schemaVersion = "1.5"
+    // Boms destination directory. Defaults to 'build/reports'
+    destination = file("build/reports")
+    // The file name for the generated BOMs (before the file format suffix). Defaults to 'bom'
+    outputName = "bom"
+    // The file format generated, can be xml, json or all for generating both. Defaults to 'all'
+    outputFormat = "json"
+    includeConfigs = ["runtimeClasspath"]
+    // Exclude test components. This list needs to be checked and, if it changed, updated for every new Kafka version.
+    // The list can be obtained by running `gradle projects | grep upgrade-system-tests`
+    skipProjects = [
+      'upgrade-system-tests-0100',
+      'upgrade-system-tests-0101',
+      'upgrade-system-tests-0102',
+      'upgrade-system-tests-0110',
+      'upgrade-system-tests-10',
+      'upgrade-system-tests-11',
+      'upgrade-system-tests-20',
+      'upgrade-system-tests-21',
+      'upgrade-system-tests-22',
+      'upgrade-system-tests-23',
+      'upgrade-system-tests-24',
+      'upgrade-system-tests-25',
+      'upgrade-system-tests-26',
+      'upgrade-system-tests-27',
+      'upgrade-system-tests-28',
+      'upgrade-system-tests-30',
+      'upgrade-system-tests-31',
+      'upgrade-system-tests-32',
+      'upgrade-system-tests-33',
+      'upgrade-system-tests-34',
+      'upgrade-system-tests-35',
+      'upgrade-system-tests-36',
+      'upgrade-system-tests-37'
+    ]
 }
 
 ext {
@@ -1011,7 +1053,7 @@ project(':core') {
     testImplementation libs.junitJupiter
     testImplementation libs.slf4jlog4j
     testImplementation libs.caffeine
-    
+
     generator project(':generator')
   }
 
